pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,302 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixManager;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.controller.rebalancer.strategy.AutoRebalanceStrategy;
+import org.apache.helix.controller.stages.ClusterDataCache;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Basic rebalance segments strategy, which rebalances offline segments using autorebalance strategy,
+ * and consuming segments using the partition assignment strategy for the table
+ */
+public class DefaultRebalanceSegmentStrategy extends BaseRebalanceSegmentStrategy {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultRebalanceSegmentStrategy.class);
+
+  protected static final String DEFAULT_DRY_RUN = "true";
+  protected static final String DEFAULT_REBALANCE_CONSUMING = "false";
+
+  private HelixManager _helixManager;
+  private HelixAdmin _helixAdmin;
+  private String _helixClusterName;
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+  public DefaultRebalanceSegmentStrategy(HelixManager helixManager) {
+    super(helixManager);
+    _helixManager = helixManager;
+    _helixAdmin = helixManager.getClusterManagmentTool();
+    _helixClusterName = helixManager.getClusterName();
+    _propertyStore = helixManager.getHelixPropertyStore();
+  }
+
+  @Override
+  public PartitionAssignment rebalancePartitionAssignment(IdealState idealState, TableConfig tableConfig,
+      RebalanceUserParams rebalanceUserParams) {
+    PartitionAssignment partitionAssignment = null;
+
+    boolean rebalanceConsuming = Boolean.valueOf(
+        rebalanceUserParams.getConfig(RebalanceUserParamConstants.REBALANCE_CONSUMING,
+            DEFAULT_REBALANCE_CONSUMING));
+    if (rebalanceConsuming) {
+      LOGGER.info("Rebalancing partition assignment for table {}", tableConfig.getTableName());
+
+      if (tableConfig.getTableType().equals(CommonConstants.Helix.TableType.REALTIME)) {
+
+        StreamPartitionAssignmentGenerator streamPartitionAssignmentGenerator = new StreamPartitionAssignmentGenerator(_propertyStore);
+        TagConfig tagConfig = new TagConfig(tableConfig, _helixManager);
+        List<String> consumingInstances = _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, tagConfig.getConsumingServerTag());
+        String tableNameWithType = tableConfig.getTableName();
+        // FIXME: where should this come from? from kafka metadata stream or just read from znode how many are existing?
+        int numPartitions = getKafkaPartitionCount(tableConfig);
```

Comment:
I think you are doing correctly. This should be the same as the logic from LLC segment manager, where it fetches the partition information from Kafka. `PinotTableIdealStateBuilder.getPartitionCount` is the correct wrapper for doing that.